File: LeetCode/502.IPO/502.IPO.java
class Solution {

    class Project {
        int profit;
        int capital;

        Project(int p, int c) {
            this.profit = p;
            this.capital = c;
        }
    }

    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        PriorityQueue<Project> maxHeap = new PriorityQueue<>(
                (p1, p2) -> p2.profit - p1.profit);

        PriorityQueue<Project> minHeap = new PriorityQueue<>(
                (p1, p2) -> p1.capital - p2.capital);

        for (int i = 0; i < profits.length; i++) {
            Project p = new Project(profits[i], capital[i]);
            if (p.capital <= w) {
                maxHeap.add(p);
            } else {
                minHeap.add(p);
            }
        }

        int num = 0;
        while (num < k) {
            if (maxHeap.isEmpty()) {
                break;
            }

            w += maxHeap.remove().profit;
            num += 1;

            while (!minHeap.isEmpty() && minHeap.peek().capital <= w) {
                maxHeap.add(minHeap.remove());
            }
        }

        return w;
    }
}

File: LeetCode/502.IPO/502.IPO.swift
class Solution {
    func findMaximizedCapital(_ k: Int, _ w: Int, _ profits: [Int], _ capital: [Int]) -> Int {
        // find which projects init capital <= w,
        // so start most profit project in these projects
        let n = profits.count

        var projects = zip(capital, profits).sorted { $0.0 < $1.0 }
        var i = 0

        var heap = Heap<Int>()
        var cur = w, count = 0

        while count < k {
            while i < n && cur >= projects[i].0 {
                heap.insert(projects[i].1)
                i += 1
            }

            if let max = heap.popMax() {
                cur += max
                count += 1
            } else {
                break
            }
        }

        return cur
    }
}
